src/build_helper/src/git.rs

```diff
@@ -27,145 +28,195 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
-    git_dir: Option<&Path>,
-) -> Result<String, String> {
-    let upstream_remote = get_rust_lang_rust_remote(config, git_dir)?;
-    let branch = config.nightly_branch;
-    for upstream_master in [format!("{upstream_remote}/{branch}"), format!("origin/{branch}")] {
-        if rev_exists(&upstream_master, git_dir)? {
-            return Ok(upstream_master);
-        }
-    }
+/// Returns true if any of the passed `paths` have changed since the `base` commit.
+pub fn has_changed_since(git_dir: &Path, base: &str, paths: &[&str]) -> bool {
+    let mut git = Command::new("git");
+    git.current_dir(git_dir);
 
-    Err("Cannot find any suitable upstream master branch".to_owned())
+    git.args(["diff-index", "--quiet", base, "--"]).args(paths);
+
+    // Exit code 0 => no changes
+    // Exit code 1 => some changes were detected
+    !git.status().expect("cannot run git diff-index").success()
 }
 
-/// Finds the nearest merge commit by comparing the local `HEAD` with the upstream branch's state.
-/// To work correctly, the upstream remote must be properly configured using `git remote add <name> <url>`.
-/// In most cases `get_closest_merge_commit` is the function you are looking for as it doesn't require remote
-/// to be configured.
-fn git_upstream_merge_base(
-    config: &GitConfig<'_>,
-    git_dir: Option<&Path>,
-) -> Result<String, String> {
-    let updated_master = updated_master_branch(config, git_dir)?;
+/// Returns the latest commit that modified `target_paths`, or `None` if no such commit was found.
+/// If `author` is `Some`, only considers commits made by that author.
+fn get_latest_commit_that_modified_files(
+    git_dir: &Path,
+    target_paths: &[&str],
+    author: &str,
+) -> Result<Option<String>, String> {
     let mut git = Command::new("git");
-    if let Some(git_dir) = git_dir {
-        git.current_dir(git_dir);
+    git.current_dir(git_dir);
+
+    git.args(["rev-list", "-n1", "--first-parent", "HEAD", "--author", author]);
+
+    if !target_paths.is_empty() {
+        git.arg("--").args(target_paths);
     }
-    Ok(output_result(git.arg("merge-base").arg(&updated_master).arg("HEAD"))?.trim().to_owned())
+    let output = output_result(&mut git)?.trim().to_owned();
+    if output.is_empty() { Ok(None) } else { Ok(Some(output)) }
 }
 
-/// Searches for the nearest merge commit in the repository that also exists upstream.
+/// Returns the most recent commit found in the local history that should definitely
+/// exist upstream. We identify upstream commits by the e-mail of the commit author.
 ///
-/// It looks for the most recent commit made by the merge bot by matching the author's email
-/// address with the merge bot's email.
-pub fn get_closest_merge_commit(
+/// If `include_head` is false, the HEAD (current) commit will be ignored and only
+/// its parents will be searched. This is useful for try/auto CI, where HEAD is
+/// actually a commit made by bors, although it is not upstream yet.
```

Comment:
I'm not seeing an include_head parameter?

Reply:
Good catch, I changed it and it now takes `CiEnv` instead, I'll reword the comment.